### PR TITLE
Update sertype interfaces to support dynamic types with SHM

### LINKED ISSUE #165: Update serdata/type functions to handle sequences/strings with shared memory
### Feature

Extend shared memory to support sequences and strings

The idea is to serialize the sample directly into the shared memory if the type is not fixed for local readers.

#### Details

This will be based on the work done in https://github.com/eclipse-cyclonedds/cyclonedds/issues/941

### PR DESCRIPTION
**Pull request by sumanth-nirmal**
This PR adds support for dynamic types (linked sequences, strings both bounded and unbounded) with SHM and modifies the shared memory tests to test the same.

This is built on top of https://github.com/eclipse-cyclonedds/cyclonedds/pull/1000/ and the former needs to be merged before merging this.

Closes #165 

@eboasson @MatthiasKillat

### CODE REVIEW

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (sumanth-nirmal) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:131]: Cleanup this duplication

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (sumanth-nirmal) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:642]: Fix the key handling

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (sumanth-nirmal) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:835]: This probably is not required, remove this (dead code)

File: src/ddscxx/tests/SharedMemory.cpp
Comment (sumanth-nirmal) [inline src/ddscxx/tests/SharedMemory.cpp:244]: Update the tests to handle the serialized data in the chunk

File: src/ddscxx/tests/SharedMemory.cpp
Comment (MatthiasKillat) [inline src/ddscxx/tests/SharedMemory.cpp:417]: maybe extend the comment and add that it is only supported for fixed-size self-contained types

**Review by MatthiasKillat — COMMENTED** [review]
The mechanisms to send dynamic types appear complete and are compatible with the interface changes in the C-API. There is still cleanup required and ongoing work in the tests to ensure it sends data with iceoryx for dynamic types now.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (sumanth-nirmal) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:179]: I agree with all the above, but with the constraints here (i.e. how the constructors for the types are generated by the IDL compiler) I think we should be good here. If you agree I will propose to fix/update this in a follow-up MR.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (sumanth-nirmal) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:187]: What you described is correct, sorry probably I didn't understand the comment completely, is there something we need to fix here?

File: .travis.yml
Comment (MatthiasKillat) [inline .travis.yml:199]: Should be set back to master or he origin repo once the branch is merged there.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (eboasson) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:761]: I added a comment to the PR on the core repository :)

File: .travis.yml
Comment (eboasson) [reply .travis.yml:199]: Yep.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (sumanth-nirmal) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:761]: I will return max here to be treated as error

File: .azure/templates/build-test.yml
Comment (sumanth-nirmal) [reply .azure/templates/build-test.yml:107]: I added this to trigger CI with the branch [`ApexAI/update-serdata-functions-to-handle-sequences-with-shared-memory`](https://github.com/ApexAI/cyclonedds/tree/update-serdata-functions-to-handle-sequences-with-shared-memory), but will drop the commit once the earlier is merged.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (MatthiasKillat) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:182]: Yes, it depends on what the deserialization expects. A constructed object or raw memory. I think in this case it is the former and so we even need to construct `T`. Note that ideally I would prefer to leave this to the deserialization (and just give it memory) because it can optimize it better (since the deserialization code alone knows what has to be done).

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp
Comment (MatthiasKillat) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:190]: No, nothing to fix.

File: src/ddscxx/tests/SharedMemory.cpp
Comment (MatthiasKillat) [inline src/ddscxx/tests/SharedMemory.cpp:70]: This could/should probably be `static` to create it only once. Otherwise we always init a rng to just get one random number from it... Probably not called often though and the actual (pseudo) randomness properties are not relevant.

**Review by MatthiasKillat — COMMENTED** [review]
Looks fine and works locally for me. The branch references to CycloneDDS need to be changed.

**Comment by sumanth-nirmal** [comment]
The CI is green with branch [`ApexAI/update-serdata-functions-to-handle-sequences-with-shared-memory`](https://github.com/ApexAI/cyclonedds/tree/update-serdata-functions-to-handle-sequences-with-shared-memory). I will now drop the commit with changes to include this branch in CI instead of Cyclone DDS master

CC @eboasson @MatthiasKillat

**Review by eboasson — APPROVED** [review]
LGTM, thanks @sumanth-nirmal and @MatthiasKillat. I'll merge it after I merge https://github.com/eclipse-cyclonedds/cyclonedds/pull/1000

**Pull request merged**
